# ***E-Techh Hacks : Anything for learning (Hackathon)***

## **Prototype Title - Gamified : A Gentle Digital Fast**


In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

# -----------------------------
# Load and prepare your data
# -----------------------------
df = pd.read_csv('/content/Cleaned_DataSet2.csv')

df.head()

Age  Gender RelationshipStatus          Occupation Organization  \
0   21    Male  In A Relationship  University Student   University   
1   21  Female             Single  University Student   University   
2   21  Female             Single  University Student   University   
3   21  Female             Single  University Student   University   
4   21  Female             Single  University Student   University   

  SocialMedia                               SocialMediaPlatforms  \
0         Yes  Facebook, Twitter, Instagram, Youtube, Discord...   
1         Yes  Facebook, Twitter, Instagram, Youtube, Discord...   
2         Yes            Facebook, Instagram, Youtube, Pinterest   
3         Yes                                Facebook, Instagram   
4         Yes                       Facebook, Instagram, Youtube   

   NumberofSocialMediaPlatforms  HoursSpent  Frequency  Distraction  \
0                             6         3.5          5            3   
1                             6         6.0          4            3   
2                             4         3.5          2            2   
3                             2         6.0          3            2   
4                             3         3.5          4            5   

   Restlessness  Anxiety  ConcentrationDifficulty  SelfComparision  \
0             2        2                        5                2   
1             2        5                        4                5   
2             1        5                        4                3   
3             1        5                        3                5   
4             4        5                        5                3   

   PostSentiment  ValidationSeeking  Depression  ActivityInterestVariance  \
0              3                  2           5                         4   
1              1                  1           5                         4   
2              3                  1           4                         2   
3              1                  2           4                         3   
4              3                  3           4                         4   

   Sleeplessness  
0              5  
1              5  
2              5  
3              2  
4              1

In [ ]:
df["User_ID"] = df.index + 1   # start from 1 instead of 0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 479 entries, 0 to 478
Data columns (total 21 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Age                           479 non-null    int64  
 1   Gender                        479 non-null    object 
 2   RelationshipStatus            479 non-null    object 
 3   Occupation                    479 non-null    object 
 4   Organization                  479 non-null    object 
 5   SocialMedia                   479 non-null    object 
 6   SocialMediaPlatforms          479 non-null    object 
 7   NumberofSocialMediaPlatforms  479 non-null    int64  
 8   HoursSpent                    479 non-null    float64
 9   Frequency                     479 non-null    int64  
 10  Distraction                   479 non-null    int64  
 11  Restlessness                  479 non-null    int64  
 12  Anxiety                       479 non-null    int64  
 13  Conce

In [ ]:
df.HoursSpent.value_counts()

HoursSpent
3.5    260
6.0    115
1.0    104
Name: count, dtype: int64

In [ ]:
df.Gender.value_counts()

Gender
Female        263
Male          211
Non Binary      4
Trans           1
Name: count, dtype: int64

In [ ]:
df.RelationshipStatus.value_counts()

RelationshipStatus
Single               284
Married              101
In A Relationship     87
Divorced               7
Name: count, dtype: int64

In [ ]:
df.Organization.value_counts()

Organization
University    308
Company       108
School         54
Government      9
Name: count, dtype: int64

In [ ]:
df['Occupation'].value_counts()

Occupation
University Student    292
Salaried Worker       131
School Student         48
Retired                 8
Name: count, dtype: int64

In [ ]:
df.SocialMediaPlatforms.value_counts()

SocialMediaPlatforms
Facebook, Instagram, Youtube                                        35
Facebook, Youtube                                                   30
Facebook, Instagram, Youtube, Snapchat                              27
Facebook, Instagram, Youtube, Snapchat, Discord                     19
Facebook, Instagram, Youtube, Discord                               18
                                                                    ..
Facebook, Instagram, Youtube, Reddit, Tiktok                         1
Facebook, Twitter, Youtube, Pinterest                                1
Instagram, Discord                                                   1
Twitter, Instagram, Youtube, Discord, Pinterest                      1
Instagram, Youtube, Snapchat, Discord, Reddit, Pinterest, Tiktok     1
Name: count, Length: 125, dtype: int64

In [ ]:
df['HoursSpent'] = df['HoursSpent'].map({
    "Less than 2": 1.0,
    "2 to 5": 3.5,
    "More than 5": 6
})

In [ ]:
df['Gender'] = df['Gender'].map({
    "Male": 0,
    "Female": 1,
    "Non binary": 2,
    "Trans": 3
})

In [ ]:
df['RelationshipStatus'] = df['RelationshipStatus'].map({
    "Single": 0,
    "Married": 1,
    "In a relationship": 2,
    "Divorced": 3
})

In [ ]:
df['Organization'] = df['Organization'].map({
    "University": 0,
    "Company": 1,
    "School": 2,
    "Government": 3
})

In [ ]:
df['Occupation'] = df['Occupation'].map({
    "University Student": 0,
    "Salaried Worker": 1,
    "School Student": 2,
    "Retired": 3

})

In [ ]:
#mode_value = df['HoursSpent'].mode()[0]
#df['HoursSpent'].fillna(mode_value, inplace=True)

/tmp/ipython-input-561323238.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [ ]:
df.isnull().sum()

Age                             0
Gender                          0
RelationshipStatus              0
Occupation                      0
Organization                    0
SocialMedia                     0
SocialMediaPlatforms            0
NumberofSocialMediaPlatforms    0
HoursSpent                      0
Frequency                       0
Distraction                     0
Restlessness                    0
Anxiety                         0
ConcentrationDifficulty         0
SelfComparision                 0
PostSentiment                   0
ValidationSeeking               0
Depression                      0
ActivityInterestVariance        0
Sleeplessness                   0
User_ID                         0
dtype: int64

In [ ]:
# Select features for clustering
features = df[["HoursSpent",
    "Distraction", "Restlessness", "Anxiety", "ConcentrationDifficulty",
    "SelfComparision", "ValidationSeeking", "Depression",
    "Sleeplessness", "ActivityInterestVariance"]]

# -----------------------------
# Scale features — very important for K-Means
# -----------------------------
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)

In [ ]:
# -----------------------------
# Find optimal k with Elbow Method
# -----------------------------
inertia = []
k_range = range(2, 10)

for k in k_range:
    kmeans = KMeans(n_clusters=k, random_state=42)
    kmeans.fit(features_scaled)
    inertia.append(kmeans.inertia_)  # Distortion / SSE

import plotly.graph_objects as go

fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=list(k_range),
        y=inertia,
        mode='lines+markers',
        marker=dict(color='royalblue', size=8),
        line=dict(width=2),
        name='Inertia'
    )
)

fig.update_layout(
    title='Elbow Method For Optimal k',
    xaxis_title='Number of clusters (k)',
    yaxis_title='Inertia (SSE)',
    xaxis=dict(tickmode='linear'),
    template='plotly_white',
    width=800,
    height=500
)

fig.show()


In [ ]:
# Suppose you choose k= from the elbow
optimal_k = 5

kmeans_final = KMeans(n_clusters=optimal_k, random_state=42)
df['Social_Impact_Type'] = kmeans_final.fit_predict(features_scaled)

# Select columns by passing a list of column names
print(df.groupby('Social_Impact_Type')[features.columns].mean())

                    HoursSpent  Distraction  Restlessness   Anxiety  \
Social_Impact_Type                                                    
0                     4.779070     4.186047      3.767442  4.348837   
1                     3.642857     3.600000      2.733333  2.790476   
2                     4.311404     4.131579      3.228070  4.438596   
3                     2.632653     2.561224      1.642857  3.979592   
4                     2.118421     1.684211      1.276316  1.921053   

                    ConcentrationDifficulty  SelfComparision  \
Social_Impact_Type                                             
0                                  4.279070         2.941860   
1                                  2.685714         2.247619   
2                                  4.166667         4.184211   
3                                  3.285714         2.632653   
4                                  1.407895         1.736842   

                    ValidationSeeking  Depression  Sl

In [ ]:
#cluster_avg = df.groupby("Social_Impact_Type")[features.columns].mean().sort_values(by="ConcentrationDifficulty")
#ordered_clusters = cluster_avg.index.tolist()

In [ ]:
cluster_avg = (
    df.groupby("Social_Impact_Type")[features.columns]
      .mean()
      .sort_values(by=["ConcentrationDifficulty", "HoursSpent"])
)
ordered_clusters = cluster_avg.index.tolist()

In [ ]:
ordered_clusters

[4, 1, 3, 2, 0]

In [ ]:
# Make a mapping
segment_names = ["Healthy", "Mild Risk","Moderate Risk","High Risk", "Critical Risk",]
cluster_to_label = {cluster: segment_names[i] for i, cluster in enumerate(ordered_clusters)}

# Apply mapping
df["Social_Impact_Label"] = df["Social_Impact_Type"].map(cluster_to_label)

print(df[["User_ID","Social_Impact_Type", "Social_Impact_Label"]].head())
print('--' * 22)
print(df[["Social_Impact_Type", "Social_Impact_Label"]].value_counts())

   User_ID  Social_Impact_Type Social_Impact_Label
0        1                   3       Moderate Risk
1        2                   0       Critical Risk
2        3                   3       Moderate Risk
3        4                   3       Moderate Risk
4        5                   0       Critical Risk
--------------------------------------------
Social_Impact_Type  Social_Impact_Label
2                   High Risk              114
1                   Mild Risk              105
3                   Moderate Risk           98
0                   Critical Risk           86
4                   Healthy                 76
Name: count, dtype: int64


In [ ]:
import plotly.express as px


counts = df['Social_Impact_Label'].value_counts().reset_index()
counts.columns = ['Social_Impact_Label', 'Count']
counts['Social_Impact_Label'] = counts['Social_Impact_Label'].astype(str)  # 👈 ensure it's treated as categorical

# Plot
fig = px.bar(
    counts,
    x='Social_Impact_Label',
    y='Count',
    text='Count',
    color='Social_Impact_Label',
    color_discrete_map={
        'Healthy': 'lightseagreen',
        'Mild Risk': 'royalblue',
        'Moderate Risk': 'mediumorchid',
        'High Risk': 'orangered',
        'Critical Risk': 'darkred'
    },  # ✅ this works now!
    title='Social Impact Labels Distribution'
)

fig.update_layout(
    width=800,
    height=500,
    xaxis_title='Social Impact Labels',
    yaxis_title='Number of Users',
    bargap=0.3
)

fig.update_traces(
    textposition='outside'
)

fig.show()


In [ ]:
df_numeric = df.select_dtypes(include=['number'])
corr = df_numeric.corr()

import plotly.express as px

fig = px.imshow(
    corr,
    text_auto=".2f",
    color_continuous_scale='RdBu_r',
    title="Correlation Heatmap (Numeric Columns Only)"
)

fig.update_layout(
    width=1000,
    height=800
)

fig.show()


In [ ]:
df.columns

Index(['Age', 'Gender', 'RelationshipStatus', 'Occupation', 'Organization',
       'SocialMedia', 'SocialMediaPlatforms', 'NumberofSocialMediaPlatforms',
       'HoursSpent', 'Frequency', 'Distraction', 'Restlessness', 'Anxiety',
       'ConcentrationDifficulty', 'SelfComparision', 'PostSentiment',
       'ValidationSeeking', 'Depression', 'ActivityInterestVariance',
       'Sleeplessness', 'User_ID', 'Social_Impact_Type',
       'Social_Impact_Label'],
      dtype='object')

In [ ]:
corr_features = [
    'User_ID',
    'Age', # Corrected column name
    'NumberofSocialMediaPlatforms',
    'HoursSpent',
    'Occupation',
    'Organization',
    'PostSentiment',
    'Frequency',
    'Distraction',
    'Restlessness',
    'Anxiety',
    'ConcentrationDifficulty',
    'SelfComparision',
    'ValidationSeeking',
    'Depression',
    'ActivityInterestVariance',
    'Sleeplessness',
    'Social_Impact_Type'

]

import plotly.express as px
import plotly.graph_objects as go

# Calculate correlation matrix again
corr_matrix = df[corr_features].corr()

# Select only correlations with target column
target_corr = corr_matrix[['Social_Impact_Type']].drop(index='Social_Impact_Type').reset_index()
target_corr.columns = ['Feature', 'Correlation']
target_corr['AbsCorrelation'] = target_corr['Correlation'].abs()

# Sort by absolute correlation
target_corr = target_corr.sort_values(by='AbsCorrelation', ascending=False)

fig = px.bar(
    target_corr,
    x='Feature',
    y='Correlation',
    color='Correlation',
    title='Feature Correlations with Social Impact Types',
    color_continuous_scale='earth',
    text=target_corr['Correlation'].round(3)
)

fig.update_layout(
    width=900,
    height=650,
    xaxis_title='Feature',
    yaxis_title='Correlation with Social Impact Types',
    bargap=0.3
)

fig.update_traces(
    textposition='outside'
)

fig.show()

In [ ]:
# ✅ Example
selected_features = [
    'User_ID',
    'Age', # Corrected column name
    'NumberofSocialMediaPlatforms',
    'HoursSpent',
    'Occupation',
    'Organization',
    'PostSentiment',
    'Frequency',
    'Distraction',
    'Restlessness',
    'Anxiety',
    'ConcentrationDifficulty',
    'SelfComparision',
    'ValidationSeeking',
    'Depression',
    'ActivityInterestVariance',
    'Sleeplessness'
]

# Create a new DataFrame for X (features)
X = df[selected_features]

# Target variable
y = df['Social_Impact_Type']


In [ ]:
from sklearn.model_selection import train_test_split

# Split without SMOTE
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,

)


In [ ]:

from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix,ConfusionMatrixDisplay
import plotly.figure_factory as ff


# 2. HistGradientBoosting Classifier
hgb = HistGradientBoostingClassifier(random_state=42)
hgb.fit(X_train, y_train)
y_pred_hgb = hgb.predict(X_test)
print('----' * 18)
print(f"📊 HistGradientBoosting Accuracy: {accuracy_score(y_test, y_pred_hgb):.2f}")
print('----' * 18)

# Confusion matrix
cm = confusion_matrix(y_test, y_pred_hgb)

# Plot confusion matrix using Plotly
labels = ["Healthy", "Mild Risk","Moderate Risk","Critical Risk", "High Risk",]
z_text = [[str(y) for y in x] for x in cm]

fig = ff.create_annotated_heatmap(
    z=cm,
    x=labels,
    y=labels,
    annotation_text=z_text,
    colorscale='Blues',
    showscale=True
)

fig.update_layout(
    title_text='Confusion Matrix - HistGradientBoosting Classifier',
    width=550,
    height=500
)

fig['data'][0]['showscale'] = True
fig.show()


------------------------------------------------------------------------
📊 HistGradientBoosting Accuracy: 0.92
------------------------------------------------------------------------


In [ ]:
X_test['Predicted_Social_Impact_Segmt'] = y_pred_hgb

In [ ]:

HoursSpent_map = {
    1.0: 'Less than 2',
    3.5: '2 to 5',
    6: 'More than 5'

}


Occupation_map = {
    0: 'University Student',
    1: 'Salaried Worker',
    2: 'School Student',
    3: 'Retired'
}

Organization_map = {
    0: 'University',
    1: 'Company',
    2: 'School',
    3: 'Government'
}

Predicted_Social_Impact_Segmt_map = {
    4: 'Healthy',
    1: 'Mild Risk',
    3: 'Moderate Risk',
    2: 'High Risk',
    0: 'Critical Risk'
}
# Apply maps

X_test['Predicted_Social_Impact_Segmt'] = X_test['Predicted_Social_Impact_Segmt'].map(Predicted_Social_Impact_Segmt_map)
X_test['Organization'] = X_test['Organization'].map(Organization_map)
X_test['Occupation'] = X_test['Occupation'].map(Occupation_map)
X_test['HoursSpent'] = X_test['HoursSpent'].map(HoursSpent_map)


In [ ]:
X_test

User_ID  Age  NumberofSocialMediaPlatforms   HoursSpent  \
93        94   30                             1       2 to 5   
320      321   21                             6  More than 5   
55        56   27                             7  More than 5   
30        31   22                             4       2 to 5   
449      450   28                             5       2 to 5   
..       ...  ...                           ...          ...   
249      250   34                             1  Less than 2   
393      394   27                             4  Less than 2   
425      426   29                             4  More than 5   
42        43   56                             1  Less than 2   
228      229   21                             3  More than 5   

             Occupation Organization  PostSentiment  Frequency  Distraction  \
93      Salaried Worker   University              3          2            5   
320  University Student   University              3          2            3   
55   University Student   University              3          4            5   
30   University Student   University              2          3            4   
449  University Student   University              3          4            3   
..                  ...          ...            ...        ...          ...   
249     Salaried Worker       School              1          1            2   
393     Salaried Worker      Company              3          2            2   
425     Salaried Worker      Company              3          4            5   
42              Retired   Government              3          1            1   
228  University Student   University              3          4            5   

     Restlessness  Anxiety  ConcentrationDifficulty  SelfComparision  \
93              3        4                        4                3   
320             2        4                        4                4   
55              5        4                        5                3   
30              3        5                        4                5   
449             1        5                        4                1   
..            ...      ...                      ...              ...   
249             1        1                        1                2   
393             1        3                        1                2   
425             5        5                        4                3   
42              1        3                        1                1   
228             3        4                        4                3   

     ValidationSeeking  Depression  ActivityInterestVariance  Sleeplessness  \
93                   3           4                         4              4   
320                  3           4                         2              4   
55                   4           5                         2              3   
30                   3           5                         4              3   
449                  1           4                         1              5   
..                 ...         ...                       ...            ...   
249                  1           1                         1              1   
393                  4           2                         2              1   
425                  4           4                         4              3   
42                   1           2                         3              1   
228                  4           4                         4              5   

    Predicted_Social_Impact_Segmt  
93                      High Risk  
320                     High Risk  
55                  Critical Risk  
30                      High Risk  
449                 Moderate Risk  
..                            ...  
249                       Healthy  
393                       Healthy  
425                     High Risk  
42                        Healthy  
228                     High Risk  

[96 rows x 18 columns]

2                   High Risk              114

1                   Mild Risk              105

3                   Moderate Risk           98

0                   Critical Risk           86

4                   Healthy                 76

In [ ]:
import plotly.express as px

# --------------------------------------------
# 1️⃣ Group your data
# --------------------------------------------
df_sunburst = (
    X_test.groupby(['HoursSpent', 'Predicted_Social_Impact_Segmt'])
    .agg({'User_ID': 'count'})
    .reset_index()
    .rename(columns={'User_ID': 'Count'})
)

#df_sunburst['Predicted_Churn_Label'] = df_sunburst['Predicted_Churn'].replace({1: 'Churn', 0: 'No Churn'})

# --------------------------------------------
# 2️⃣ Build sunburst
# --------------------------------------------
fig = px.sunburst(
    df_sunburst,
    path=['HoursSpent', 'Predicted_Social_Impact_Segmt'],  # hierarchy: parent → child
    values='Count',
    color='HoursSpent',  # color by CLV Cluster!
    color_discrete_sequence=px.colors.sequential.Darkmint,
    title="Sunburst: Social Impact Segments with their Hours Spent"
)

# --------------------------------------------
# 3️⃣ Style
# --------------------------------------------
fig.update_traces(
    textinfo='label+percent entry+value'
)

fig.update_layout(
    width=800,
    height=600
)

fig.show()


In [ ]:
# Filter rows where Predicted_Churn is 'Churn'
df_2_to_5_rows = X_test[X_test['HoursSpent'] == '2 to 5']

# Select only individual_id and CLV_Cluster columns
case1_info = df_2_to_5_rows[['User_ID', 'Predicted_Social_Impact_Segmt']]

# Display result
print(case1_info.head(12).to_markdown(index=False, numalign="left", stralign="left"))


| User_ID   | Predicted_Social_Impact_Segmt   |
|:----------|:--------------------------------|
| 94        | High Risk                       |
| 31        | High Risk                       |
| 450       | Moderate Risk                   |
| 457       | Mild Risk                       |
| 434       | High Risk                       |
| 77        | Mild Risk                       |
| 272       | Mild Risk                       |
| 239       | High Risk                       |
| 465       | Moderate Risk                   |
| 105       | Mild Risk                       |
| 281       | High Risk                       |
| 279       | High Risk                       |


In [ ]:
# Filter rows where Predicted_Churn is 'Churn'
df_more_than_5_rows = X_test[X_test['HoursSpent'] == 'More than 5']

# Select only individual_id and CLV_Cluster columns
case2_info = df_more_than_5_rows[['User_ID', 'Predicted_Social_Impact_Segmt']]

# Display result
print(case2_info.head(12).to_markdown(index=False, numalign="left", stralign="left"))


| User_ID   | Predicted_Social_Impact_Segmt   |
|:----------|:--------------------------------|
| 321       | High Risk                       |
| 56        | Critical Risk                   |
| 83        | Critical Risk                   |
| 358       | Critical Risk                   |
| 322       | Mild Risk                       |
| 230       | Moderate Risk                   |
| 437       | Mild Risk                       |
| 378       | Mild Risk                       |
| 401       | Critical Risk                   |
| 395       | Critical Risk                   |
| 73        | High Risk                       |
| 442       | High Risk                       |


In [ ]:
# Filter rows where Predicted_Churn is 'Churn'
df_Less_than_2_rows = X_test[X_test['HoursSpent'] == 'Less than 2']

# Select only individual_id and CLV_Cluster columns
case3_info = df_Less_than_2_rows[['User_ID', 'Predicted_Social_Impact_Segmt']]

# Display result
print(case3_info.head(12).to_markdown(index=False, numalign="left", stralign="left"))


| User_ID   | Predicted_Social_Impact_Segmt   |
|:----------|:--------------------------------|
| 10        | Healthy                         |
| 298       | Moderate Risk                   |
| 420       | Mild Risk                       |
| 64        | Moderate Risk                   |
| 182       | Healthy                         |
| 471       | Mild Risk                       |
| 219       | Moderate Risk                   |
| 200       | Moderate Risk                   |
| 154       | Moderate Risk                   |
| 19        | Moderate Risk                   |
| 156       | Moderate Risk                   |
| 339       | Moderate Risk                   |


In [ ]:
print(case1_info['Predicted_Social_Impact_Segmt'].value_counts())


Predicted_Social_Impact_Segmt
Mild Risk        14
High Risk        12
Moderate Risk     9
Critical Risk     7
Healthy           1
Name: count, dtype: int64


In [ ]:
print(case2_info['Predicted_Social_Impact_Segmt'].value_counts())

Predicted_Social_Impact_Segmt
High Risk        10
Critical Risk     8
Mild Risk         4
Moderate Risk     3
Healthy           1
Name: count, dtype: int64


In [ ]:
print(case3_info['Predicted_Social_Impact_Segmt'].value_counts())

Predicted_Social_Impact_Segmt
Moderate Risk    14
Healthy           9
Mild Risk         4
Name: count, dtype: int64


In [ ]:
X_test.to_csv('Social Impact Test.csv', index=False)

In [ ]:
import pandas as pd

# Load the CSV file into a DataFrame

X_test_df = pd.read_csv('Social Impact Test.csv')

In [ ]:
!pip install txtai[pipeline]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.6/41.6 kB 3.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.7/50.7 MB 19.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.5/48.5 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.6/162.6 kB 13.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Using cached ninja-1.13.0-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (5.1 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.3/61.3 MB 12.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 999.8/999.8 kB 57.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 224.9/224.9 kB 17.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.4/31.4 MB 63.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
!pip install langchain

In [ ]:
!pip install langchain-community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 29.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.0 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.32.4, but you have requests 2.32.5 which is incompatible.


In [ ]:

from txtai.pipeline import LLM as TxtaiLLM

#txtai_llm = TxtaiLLM("MaziyarPanahi/gemma-2-2b-it-GGUF/gemma-2-2b-it.Q8_0.gguf")

txtai_llm = TxtaiLLM("muranAI/gemma-3n-E4B-it-GGUF/gemma-3n-e4b-it-q2_k.gguf")

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package cmudict to /root/nltk_data...
[nltk_data]   Unzipping corpora/cmudict.zip.
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


gemma-3n-e4b-it-q2_k.gguf:   0%|          | 0.00/2.76G [00:00<?, ?B/s]

llama_kv_cache_unified_iswa: using full-size SWA cache (ref: https://github.com/ggml-org/llama.cpp/pull/13194#issuecomment-2868343055)


In [ ]:
# =========================
# Gamified Digital Wellness Recommender
# =========================

import pandas as pd
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from langchain.schema import StrOutputParser
from txtai.pipeline import LLM as TxtaiLLM
from langchain.llms.base import LLM as LangChainLLM

# =========================
# Wrap txtai LLM for LangChain
# =========================
txtai_llm_instance = TxtaiLLM("muranAI/gemma-3n-E4B-it-GGUF/gemma-3n-e4b-it-q2_k.gguf")

class TxtaiLangChainLLM(LangChainLLM):
    def _call(self, prompt, stop=None):
        return txtai_llm_instance(prompt)

    @property
    def _llm_type(self):
        return "txtai"

llm = TxtaiLangChainLLM()

# =========================
# Helper: Map HoursSpent → Buckets
# =========================
def map_hours_to_bucket(hours):
    """
    Convert numeric HoursSpent into exact rule-based buckets.
    """
    try:
        hours = float(hours)
        if hours > 5:
            return "More than 5"
        elif 2 <= hours <= 5:
            return "2 to 5"
        else:
            return "Less than 2"
    except:
        # In case HoursSpent is already categorical string
        val = str(hours).strip()
        if val in ["More than 5", "2 to 5", "Less than 2"]:
            return val
        else:
            return "Unknown"

# =========================
# Recommendation Prompt (Wellness + Gamified Nudges)
# =========================
recommend_action_prompt = PromptTemplate.from_template("""
You are a digital wellness mentor who gives gamified nudges and challenges.

RULES:
1. If Predicted_Social_Impact_Segmt == 'Critical Risk':
   → 🚨 Level: SOS Guardian → Challenge: Start with a "1-hour no-screen" mission today, then connect with a supportive friend or family member.

2. If Predicted_Social_Impact_Segmt == 'High Risk':
   → ⚖️ Level: Balance Builder → Challenge: Try 2 offline activities today (walk, journaling, outdoor play) and track how you feel.

3. If Predicted_Social_Impact_Segmt == 'Moderate Risk':
   → 🛠️ Level: Habit Shaper → Challenge: Set a 2-hour screen-time cap and swap one digital activity with a hobby.

4. If Predicted_Social_Impact_Segmt == 'Mild Risk':
   → 🌱 Level: Growth Seeker → Challenge: Share 1 positive post, call a family member, or join a community activity.

5. If Predicted_Social_Impact_Segmt == 'Healthy':
   → 🏆 Level: Wellness Champion → Challenge: Celebrate your balance streak, log your digital habits, and mentor a peer with your success tips.

HOURSPENT LEVELS:
- If HoursSpent == 'More than 5':
   → 🔴 Time Overlord → Extra Challenge: Take a 5-min break every hour and spend at least 30 mins outdoors.

- If HoursSpent == '2 to 5':
   → 🟡 Balance Keeper → Bonus: Maintain your healthy range by mixing short offline breaks with creative hobbies.

- If HoursSpent == 'Less than 2':
   → 🟢 Focus Master → Bonus: You’re on a great streak! Keep celebrating your strong balance and inspire others.

OUTPUT FORMAT:
Return ONLY the matched recommendation lines.
- Do NOT repeat or restate the rules.
- Do NOT include conditions or "If..." statements.
- Only output the selected challenge(s).


INPUT:
- Social Impact Segment: {Predicted_Social_Impact_Segmt}
- HoursSpent: {HoursSpent}
""")


recommend_chain = LLMChain(
    llm=llm,
    prompt=recommend_action_prompt,
    output_parser=StrOutputParser()
)

# =========================
# Short Explanation Prompt
# =========================
llm_explain_prompt = PromptTemplate.from_template("""
You are a friendly digital well-being mentor.

User Info:
- Social Impact Segment: {Predicted_Social_Impact_Segmt}
- HoursSpent: {HoursSpent}

System's recommended action:
{action}

In 1–2 upbeat sentences (max 30 words), explain why this challenge fits the user.
Keep it encouraging and gamified.

Answer:
""")

llm_explain_chain = LLMChain(
    llm=llm,
    prompt=llm_explain_prompt,
    output_parser=StrOutputParser()
)

# =========================
# Run recommender for X_test with User Input
# =========================
def run_recommender_for_user(X_test_df):
    try:
        # Ask user to input User_ID
        user_id = int(input("Enter User_ID: "))

        row = X_test_df.loc[X_test["User_ID"] == user_id]

        if row.empty:
            return f"❌ User_ID {user_id} not found in dataset."

        row = row.iloc[0]
        seg = str(row["Predicted_Social_Impact_Segmt"]).strip()
        hrs_bucket = map_hours_to_bucket(row["HoursSpent"])

        # Generate challenge
        action = recommend_chain.run({
            "Predicted_Social_Impact_Segmt": seg,
            "HoursSpent": hrs_bucket
        })

        # Generate short explanation
        explanation = llm_explain_chain.run({
            "Predicted_Social_Impact_Segmt": seg,
            "HoursSpent": hrs_bucket,
            "action": action
        })

        return {
            "User_ID": row["User_ID"],
            "Segment": seg,
            "HoursSpent": hrs_bucket,
            "Challenge": action,
            "Why": explanation
        }

    except ValueError:
        return "❌ Invalid input. Please enter a numeric User_ID."


# =========================
# Example usage
# =========================
if __name__ == "__main__":
    result = run_recommender_for_user(X_test)
    if isinstance(result, dict):
        print("\n🎯 Personalized Recommendation")
        print(f"User_ID: {result['User_ID']}")
        print(f"Segment: {result['Segment']}")
        print(f"HoursSpent: {result['HoursSpent']}")
        print(f"Challenge: {result['Challenge']}")
        print(f"Why: {result['Why']}")
    else:
        print(result)


llama_kv_cache_unified_iswa: using full-size SWA cache (ref: https://github.com/ggml-org/llama.cpp/pull/13194#issuecomment-2868343055)


Enter User_ID: 321

🎯 Personalized Recommendation
User_ID: 321
Segment: High Risk
HoursSpent: More than 5
Challenge: ```
🚨 Level: Balance Builder → Challenge: Try 2 offline activities today (walk, journaling, outdoor play) and track how you feel.
    Bonus: 🔴 Time Overlord → Extra Challenge: Take a 5-min break every hour and spend at least 30 mins outdoors.
```
Why: This challenge is perfect for you! Since you're navigating a high-impact area, a balance builder will help you reconnect with yourself and the world outside. The bonus is all about nurturing your connection with the outdoors, which often has a calming effect on the mind.


In [ ]:
import gradio as gr
import pandas as pd
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from langchain.schema import StrOutputParser
from txtai.pipeline import LLM as TxtaiLLM
from langchain.llms.base import LLM as LangChainLLM



# =========================
# Wrap txtai LLM
# =========================
txtai_llm_instance = TxtaiLLM("muranAI/gemma-3n-E4B-it-GGUF/gemma-3n-e4b-it-q2_k.gguf")

class TxtaiLangChainLLM(LangChainLLM):
    def _call(self, prompt, stop=None):
        return txtai_llm_instance(prompt)

    @property
    def _llm_type(self):
        return "txtai"

llm = TxtaiLangChainLLM()

# =========================
# Recommendation Prompt
# =========================
recommend_action_prompt = PromptTemplate.from_template("""
You are a digital wellness mentor who gives gamified nudges and challenges.

RULES:
1. If Predicted_Social_Impact_Segmt == 'Critical Risk':
   → 🚨 Level: SOS Guardian → Challenge: Start with a "1-hour no-screen" mission today, then connect with a supportive friend or family member.

2. If Predicted_Social_Impact_Segmt == 'High Risk':
   → ⚖️ Level: Balance Builder → Challenge: Try 2 offline activities today (walk, journaling, outdoor play) and track how you feel.

3. If Predicted_Social_Impact_Segmt == 'Moderate Risk':
   → 🛠️ Level: Habit Shaper → Challenge: Set a 2-hour screen-time cap and swap one digital activity with a hobby.

4. If Predicted_Social_Impact_Segmt == 'Mild Risk':
   → 🌱 Level: Growth Seeker → Challenge: Share 1 positive post, call a family member, or join a community activity.

5. If Predicted_Social_Impact_Segmt == 'Healthy':
   → 🏆 Level: Wellness Champion → Challenge: Celebrate your balance streak, log your digital habits, and mentor a peer with your success tips.

HOURSPENT LEVELS:
- If HoursSpent == 'More than 5':
   → 🔴 Time Overlord → Extra Challenge: Take a 5-min break every hour and spend at least 30 mins outdoors.

- If HoursSpent == '2 to 5':
   → 🟡 Balance Keeper → Bonus: Maintain your healthy range by mixing short offline breaks with creative hobbies.

- If HoursSpent == 'Less than 2':
   → 🟢 Focus Master → Bonus: You’re on a great streak! Keep celebrating your strong balance and inspire others.

OUTPUT FORMAT:
Return ONLY the matched recommendation lines.
- Do NOT repeat or restate the rules.
- Do NOT include conditions or "If..." statements.
- Only output the selected challenge(s).


INPUT:
- Social Impact Segment: {Predicted_Social_Impact_Segmt}
- HoursSpent: {HoursSpent}
""")

recommend_chain = LLMChain(
    llm=llm,
    prompt=recommend_action_prompt,
    output_parser=StrOutputParser()
)

# =========================
# Explanation Prompt
# =========================
llm_explain_prompt = PromptTemplate.from_template("""
You are a friendly digital well-being mentor.

User Info:
- Social Impact Segment: {Predicted_Social_Impact_Segmt}
- HoursSpent: {HoursSpent}

System's recommended action:
{action}

In 1–2 upbeat sentences (max 30 words), explain why this challenge fits the user.
Keep it encouraging and gamified.

Answer:
""")

llm_explain_chain = LLMChain(
    llm=llm,
    prompt=llm_explain_prompt,
    output_parser=StrOutputParser()
)

# =========================
# Backend Logic
# =========================
def fetch_user_info(user_id):
    row = X_test_df[X_test_df['User_ID'] == int(user_id)]
    if row.empty:
        return "❌ User not found", "", ""

    row = row.iloc[0]
    seg = str(row["Predicted_Social_Impact_Segmt"]).strip()
    hrs_bucket = str(row["HoursSpent"]).strip()

    # Generate recommendation
    action = recommend_chain.run({
        "Predicted_Social_Impact_Segmt": seg,
        "HoursSpent": hrs_bucket
    })

    # Generate explanation
    explanation = llm_explain_chain.run({
        "Predicted_Social_Impact_Segmt": seg,
        "HoursSpent": hrs_bucket,
        "action": action
    })

    user_info = f"User ID: {row['User_ID']}\nSegment: {seg}\nHoursSpent: {hrs_bucket}"
    return user_info, action, explanation

# =========================
# Gradio UI
# =========================
with gr.Blocks() as demo:
    gr.Markdown("# 🌍💡 Gamified Digital Wellness Recommender")

    with gr.Row():
        with gr.Column(scale=1):
            user_id = gr.Dropdown(
                choices=sorted(X_test_df["User_ID"].unique().tolist()),
                label="Select User ID"
            )
            fetch_btn = gr.Button("Get Recommendation")

        with gr.Column(scale=2):
            user_info = gr.Textbox(label="User Info", lines=3)
            system_action = gr.Textbox(label="Challenge", lines=3)
            llm_explanation = gr.Textbox(label="Why This Challenge?", lines=3)

    fetch_btn.click(
        fetch_user_info,
        inputs=[user_id],
        outputs=[user_info, system_action, llm_explanation]
    )

demo.launch()


llama_kv_cache_unified_iswa: using full-size SWA cache (ref: https://github.com/ggml-org/llama.cpp/pull/13194#issuecomment-2868343055)


It looks like you are running Gradio on a hosted Jupyter notebook, which requires `share=True`. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://2c01a23b8d342f6f9f.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)
